clang/test/dpct/syncthreads.cu

```diff
@@ -149,4 +157,246 @@ __global__ void test9(S4 a) {
+
+__device__ int bar12(int num) {
+  int n = num - 1;
+  n |= n >> 1;
+  return n + 1;
+}
+
+__global__ void test12(uint8_t *pout) {
+  int idx = 456;
+  idx = bar12(idx);
+  //CHECK:  pout[idx] = 123;
+  //CHECK-NEXT:  item_ct1.barrier(sycl::access::fence_space::local_space);
+  //CHECK-NEXT:  uint8_t a = 4;
+  pout[idx] = 123;
+  __syncthreads();
+  uint8_t a = 4;
+  uint8_t b = abs(a);
+}
+
+__device__ int d_a[10];
+
+__device__ int bar13(int num) {
+  int n = num - 1;
+  n |= d_a[1] >> 1;
+  return n + 1;
+}
+
+__global__ void test13(uint8_t *pout) {
+  int idx = 456;
+  idx = bar13(idx);
+  //CHECK:  pout[idx] = 456;
+  //CHECK-NEXT:  /*
+  //CHECK-NEXT:  DPCT1065:{{[0-9]+}}: Consider replacing sycl::nd_item::barrier() with sycl::nd_item::barrier(sycl::access::fence_space::local_space) for better performance if there is no access to global memory.
+  //CHECK-NEXT:  */
+  //CHECK-NEXT:  item_ct1.barrier();
+  //CHECK-NEXT:  uint8_t a = 4;
+  pout[idx] = 456;
+  __syncthreads();
+  uint8_t a = 4;
+  uint8_t b = abs(a);
+}
+
+__device__ void barbar14() {}
+
+__device__ int bar14() {
+  barbar14();
+  int n = 123;
+  return n + 1;
+}
+
+__global__ void test14(uint8_t *pout) {
+  int idx = 456;
+  idx = bar14();
+  //CHECK:  pout[idx] = 789;
+  //CHECK-NEXT:  item_ct1.barrier(sycl::access::fence_space::local_space);
+  //CHECK-NEXT:  uint8_t a = 4;
+  pout[idx] = 789;
+  __syncthreads();
+  uint8_t a = 4;
+  uint8_t b = abs(a);
+}
+
+__global__ void test15(float *_res) {
+  __shared__ float2 ker_even[123]; // shared mem
+  //CHECK:  float *res = _res + 123;
+  //CHECK-NEXT:  res[2] = 123;
+  //CHECK-NEXT:  item_ct1.barrier(sycl::access::fence_space::local_space);
+  //CHECK-NEXT:  res += 2;
+  float *res = _res + 123;
+  res[2] = 123;
+  __syncthreads();
+  res += 2;
+}
+
+template <class T> __global__ void test16(T *res) {
+  //CHECK:  auto a = res[2];
+  //CHECK-NEXT:  item_ct1.barrier(sycl::access::fence_space::local_space);
+  //CHECK-NEXT:  a++;
+  auto a = res[2];
+  __syncthreads();
+  a++;
+}
+
+template <class T> void foo16(int grid, int block, T *res, cudaStream_t stream) {
+  test16<T><<<grid, block, 0, stream>>>(res);
+}
+
+template void foo16<half>(int grid, int block, half *res, cudaStream_t stream);
+template void foo16<float>(int grid, int block, float *res, cudaStream_t stream);
+
+__global__ void test17(float *f) {
+  __shared__ float ff[10];
+  //CHECK:  *((sycl::float2 *)&ff[2]) = *((sycl::float2 *)&f[123]);
+  //CHECK-NEXT:  item_ct1.barrier(sycl::access::fence_space::local_space);
+  item_ct1.barrier(sycl::access::fence_space::local_space);
+  *((float2 *)&ff[2]) = *((float2 *)&f[123]);
+  __syncthreads();
+}
+
+__global__ void test18(unsigned int *aaa) {
+  __shared__ unsigned int bbb[2][10];
+  //CHECK:  *((sycl::uint4 *)(&aaa[5])) = *((sycl::uint4 *)&bbb[1][2]);
+  //CHECK-NEXT:  item_ct1.barrier(sycl::access::fence_space::local_space);
+  *((uint4 *)(&aaa[5])) = *((uint4 *)&bbb[1][2]);
+  __syncthreads();
+}
```

Comment:
For write-after-write case, all write access should only before syncthreads or lonly after syncthreads

Reply:
fixed